File: house_price_frontend/frontend_phase4/src/types/prediction.ts
/**
 * TypeScript types mirroring the backend Pydantic schema
 * (backend/app/schemas/prediction.py).
 *
 * Keep these in sync with the backend — they define the exact
 * shape of the /predict request and response bodies.
 */

export type Furnishing = "Furnished" | "Semi-Furnished" | "Unfurnished";
export type Transaction = "New Property" | "Resale";
export type Ownership = "Freehold" | "Leasehold" | "Co-operative Society" | "Power of Attorney";
export type Facing =
  | "East"
  | "West"
  | "North"
  | "South"
  | "North-East"
  | "North-West"
  | "South-East"
  | "South-West";

export interface PredictionRequest {
  location: string;
  carpet_area_sqft: number;
  floor_num: number;
  bathroom: number;
  balcony: number;
  car_parking: number;
  furnishing: Furnishing | "";
  transaction: Transaction | "";
  ownership: Ownership | "";
  facing: Facing | "";
}

export interface PredictionResponse {
  predicted_price: number;
}

export interface HealthResponse {
  status: string;
}

/** Shape of the form state before submission (all fields as strings for controlled inputs). */
export interface PredictionFormState {
  location: string;
  carpet_area_sqft: string;
  floor_num: string;
  bathroom: string;
  balcony: string;
  car_parking: string;
  furnishing: Furnishing | "";
  transaction: Transaction | "";
  ownership: Ownership | "";
  facing: Facing | "";
}

export const emptyFormState: PredictionFormState = {
  location: "",
  carpet_area_sqft: "",
  floor_num: "",
  bathroom: "",
  balcony: "0",
  car_parking: "0",
  furnishing: "",
  transaction: "",
  ownership: "",
  facing: "",
};

export const FURNISHING_OPTIONS: Furnishing[] = ["Furnished", "Semi-Furnished", "Unfurnished"];
export const TRANSACTION_OPTIONS: Transaction[] = ["New Property", "Resale"];
export const OWNERSHIP_OPTIONS: Ownership[] = [
  "Freehold",
  "Leasehold",
  "Co-operative Society",
  "Power of Attorney",
];
export const FACING_OPTIONS: Facing[] = [
  "East",
  "West",
  "North",
  "South",
  "North-East",
  "North-West",
  "South-East",
  "South-West",
];

File: house_price_backend/backend/app/schemas/prediction.py
"""
Pydantic schemas for the /predict endpoint.

Field names and types mirror the exact columns the notebook's pipeline
was trained on:

    numeric_features = ["carpet_area_sqft", "floor_num", "bathroom", "balcony", "car_parking"]
    categorical_features = ["location_grouped", "Furnishing", "Transaction", "Ownership", "facing"]

`location` is mapped to `location_grouped` server-side (see services/preprocessing.py),
so unseen locations are grouped into "Other" exactly like in training.
"""
from pydantic import BaseModel, Field, field_validator


class PredictionRequest(BaseModel):
    location: str = Field(..., description="Property location / area name")
    carpet_area_sqft: float = Field(..., gt=0, description="Carpet area in square feet")
    floor_num: int = Field(..., ge=-1, description="Floor number (0 = ground, -1 = basement)")
    bathroom: int = Field(..., ge=0, description="Number of bathrooms")
    balcony: int = Field(0, ge=0, description="Number of balconies")
    car_parking: int = Field(0, ge=0, description="Number of car parking spots")
    furnishing: str = Field(..., description='"Furnished" | "Semi-Furnished" | "Unfurnished"')
    transaction: str = Field(..., description='"New Property" | "Resale"')
    ownership: str = Field(..., description="Ownership type, e.g. Freehold / Leasehold")
    facing: str = Field(..., description="Facing direction, e.g. East / North")

    @field_validator("location", "furnishing", "transaction", "ownership", "facing")
    @classmethod
    def strip_and_require_non_empty(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("must not be empty")
        return v

    model_config = {
        "json_schema_extra": {
            "example": {
                "location": "Sector 62",
                "carpet_area_sqft": 1200,
                "floor_num": 3,
                "bathroom": 2,
                "balcony": 1,
                "car_parking": 1,
                "furnishing": "Semi-Furnished",
                "transaction": "Resale",
                "ownership": "Freehold",
                "facing": "East",
            }
        }
    }


class PredictionResponse(BaseModel):
    predicted_price: float


class HealthResponse(BaseModel):
    status: str = "ok"
